src/toil/leader.py

```diff
@@ -228,262 +231,293 @@ def run(self):
 
         return self.jobStore.getRootJobReturnValue()
 
+    def _handledFailedSuccessor(self, jobNode, jobGraph, successorJobStoreID):
+        """Deal with the successor having failed. Return True if there are
+        still active successors. Return False if all successors have failed
+        and the job is queued to run to handle the failed successors."""
+        logger.debug("Successor job: %s of job: %s has failed """
+                     "predecessors", jobNode, jobGraph)
+
+        # Add the job to the set having failed successors
+        self.toilState.hasFailedSuccessors.add(jobGraph.jobStoreID)
+
+        # Reduce active successor count and remove the successor as an active successor of the job
+        self.toilState.successorCounts[jobGraph.jobStoreID] -= 1
+        assert self.toilState.successorCounts[jobGraph.jobStoreID] >= 0
+        self.toilState.successorJobStoreIDToPredecessorJobs[successorJobStoreID].remove(jobGraph)
+        if len(self.toilState.successorJobStoreIDToPredecessorJobs[successorJobStoreID]) == 0:
+            self.toilState.successorJobStoreIDToPredecessorJobs.pop(successorJobStoreID)
+
+        # If the job now has no active successors add to active jobs
+        # so it can be processed as a job with failed successors
+        if self.toilState.successorCounts[jobGraph.jobStoreID] == 0:
+            logger.debug("Job: %s has no successors to run "
+                         "and some are failed, adding to list of jobs "
+                         "with failed successors", jobGraph)
+            self.toilState.successorCounts.pop(jobGraph.jobStoreID)
+            self.toilState.updatedJobs.add((jobGraph, 0))
+            return False
+
+
+    def _checkSuccssorReadyToRunMultplePredecessors(self, jobGraph, jobNode, successorJobStoreID):
+        """Handle the special cases of checking if a successor job is
+        ready to run when there are multiple predecessors"""
+        # FIXME: it appears that the reason for special handling is to avoid
+        # recomputing the jobgraph, unclear is this is actually the case or
+        # how important it is.  If possible, it would be clearer without this
+        # special cases.
```

Comment:
I'm fairly sure we don't actually need to load the jobGraph here. We write to the `predecessorsFinished` attribute, but since the jobGraph is likely never written *back* to the job store, there are two possibilities:

- the logic is wrong, or
- `predecessorsFinished` is leader state.

It looks like it's the latter, because that attribute is only ever touched by the leader. Revising predecessorsFinished to be leader state and not job state would make this a little bit simpler.

Reply:
As for whether we could get rid of the special case: yeah, probably. Right now special-casing this is avoiding a ton of leader-blocking jobGraph loads, which is a good thing. But if those go away, I don't really see a reason to have the 1-predecessor case be special.